frame/lottery/src/lib.rs

```diff
@@ -0,0 +1,298 @@
+	// The existential deposit is not part of the pot so lottery account never gets deleted.
+	fn pot() -> (T::AccountId, BalanceOf<T>) {
+		let account_id = Self::account_id();
+		let balance = T::Currency::free_balance(&account_id)
+			.saturating_sub(T::Currency::minimum_balance());
+
+		(account_id, balance)
+	}
+
+	// Converts a vector of calls into a vector of call indices.
+	fn calls_to_indices(calls: &[<T as Config>::Call]) -> Result<Vec<CallIndex>, DispatchError> {
+		let mut indices = Vec::with_capacity(calls.len());
+		for c in calls.iter() {
+			let index = Self::call_to_index(c)?;
+			indices.push(index)
+		}
+		Ok(indices)
+	}
+
+	// Convert a call to it's call index by encoding the call and taking the first two bytes.
+	fn call_to_index(call: &<T as Config>::Call) -> Result<CallIndex, DispatchError> {
+		let encoded_call = call.encode();
+		if encoded_call.len() < 2 { Err(Error::<T>::EncodingFailed)? }
+		return Ok((encoded_call[0], encoded_call[1]))
+	}
+
+	// Logic for buying a ticket.
+	fn do_buy_ticket(caller: &T::AccountId, call: &<T as Config>::Call) -> DispatchResult {
+		// Check the call is valid lottery
+		let config = Lottery::<T>::get().ok_or(Error::<T>::NotConfigured)?;
+		let block_number = frame_system::Module::<T>::block_number();
+		ensure!(block_number >= config.start, Error::<T>::NotStarted);
+		ensure!(block_number < config.end, Error::<T>::AlreadyEnded);
+		let call_index = Self::call_to_index(call)?;
+		ensure!(config.calls.iter().any(|c| call_index == *c), Error::<T>::InvalidCall);
+		let ticket_count = TicketsCount::get();
+		let new_ticket_count = ticket_count.checked_add(1).ok_or(Error::<T>::Overflow)?;
+		// Try to update the participant status
+		Participants::<T>::try_mutate(&caller, |(lottery_index, participating_calls)| -> DispatchResult {
+			let index = LotteryIndex::get();
+			// If lottery index doesn't match, then reset participating calls and index.
+			if *lottery_index != index {
+				*participating_calls = Vec::new();
+				*lottery_index = index;
+			} else {
+				// Check that user is not already participating under this call.
+				ensure!(!participating_calls.iter().any(|c| call_index == *c), Error::<T>::AlreadyParticipating);
+			}
+			// Check user has enough funds and send it to the Lottery account.
+			T::Currency::transfer(caller, &Self::account_id(), config.price, KeepAlive)?;
+			// Create a new ticket.
+			TicketsCount::put(new_ticket_count);
+			Tickets::<T>::insert(ticket_count, caller.clone());
+			participating_calls.push(call_index);
+			Ok(())
+		})?;
+
+		Self::deposit_event(RawEvent::TicketBought(caller.clone(), call.clone()));
+
+		Ok(())
+	}
+
+	// Randomly choose a winner from among the total number of participants.
+	fn choose_winner(total: u32) -> u32 {
+		let random_seed = T::Randomness::random(b"lottery");
+		let random_number = <u32>::decode(&mut random_seed.as_ref())
+			.expect("secure hashes should always be bigger than u32; qed");
+		random_number % total
```

Comment:
https://stackoverflow.com/questions/10984974/why-do-people-say-there-is-modulo-bias-when-using-a-random-number-generator

Reply:
I don't think I can reasonably loop in the runtime for some unknown number of times.... Given large N, this should be nominal, or am I undervaluing this issue?

Reply:
Not an export on this. Someone else can judge if this is safe enough.

Reply:
if I'm correct `u32::max_value() % total` first numbers have `floor(u32::max_value() / total) + 1` chances and the other other have `floor(u32::max_value() / total)` chances.

So generally the closest `total` is to `u32` the more bias there can be.

So if my math are correct, you could do to mitigate:
* (use u64 instead of u32. But u32 may already be way higher than `total`)
* or better you could do the suggested loop but only a few time like 2 or 3 times.
```c
int x;

do {
    x = rand();
} while (x >= (RAND_MAX - RAND_MAX % n));

x %= n;
```

A line mentioning the biases should be put in the doc IMO.

Reply:
fixed